File: internal/domain/llm.go
package domain

import (
	"context"
	"encoding/json"
	"fmt"
)

// LLMTool represents a tool the AI can use
type LLMTool struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	InputSchema json.RawMessage `json:"input_schema"`
}

// LLMChatRequest represents a request to the LLM chat endpoint
type LLMChatRequest struct {
	WorkspaceID   string       `json:"workspace_id"`
	IntegrationID string       `json:"integration_id"`
	Messages      []LLMMessage `json:"messages"`
	MaxTokens     int          `json:"max_tokens,omitempty"`
	SystemPrompt  string       `json:"system_prompt,omitempty"`
	Tools         []LLMTool    `json:"tools,omitempty"`
}

// LLMMessage represents a chat message
type LLMMessage struct {
	Role    string `json:"role"` // "user" or "assistant"
	Content string `json:"content"`
}

// LLMChatEvent represents an SSE event sent during streaming
type LLMChatEvent struct {
	Type         string                 `json:"type"`                    // "text", "tool_use", "server_tool_start", "server_tool_result", "done", "error"
	Content      string                 `json:"content,omitempty"`       // Text content for "text" events
	Error        string                 `json:"error,omitempty"`         // Error message for "error" events
	ToolName     string                 `json:"tool_name,omitempty"`     // Tool name for "tool_use" events
	ToolInput    map[string]interface{} `json:"tool_input,omitempty"`    // Tool input for "tool_use" events
	InputTokens  *int64                 `json:"input_tokens,omitempty"`  // Token count (done event only)
	OutputTokens *int64                 `json:"output_tokens,omitempty"` // Token count (done event only)
	InputCost    *float64               `json:"input_cost,omitempty"`    // Cost in USD (done event only)
	OutputCost   *float64               `json:"output_cost,omitempty"`   // Cost in USD (done event only)
	TotalCost    *float64               `json:"total_cost,omitempty"`    // Total cost in USD (done event only)
	Model        string                 `json:"model,omitempty"`         // Model used (done event only)
}

// Validate validates the LLM chat request
func (r *LLMChatRequest) Validate() error {
	if r.WorkspaceID == "" {
		return fmt.Errorf("workspace_id is required")
	}
	if r.IntegrationID == "" {
		return fmt.Errorf("integration_id is required")
	}
	if len(r.Messages) == 0 {
		return fmt.Errorf("at least one message is required")
	}
	for i, msg := range r.Messages {
		if msg.Role != "user" && msg.Role != "assistant" {
			return fmt.Errorf("message %d: role must be 'user' or 'assistant'", i)
		}
		if msg.Content == "" {
			return fmt.Errorf("message %d: content is required", i)
		}
	}
	if r.MaxTokens < 0 || r.MaxTokens > 8192 {
		return fmt.Errorf("max_tokens must be between 0 and 8192")
	}
	return nil
}

//go:generate mockgen -destination mocks/mock_llm_service.go -package mocks github.com/Notifuse/notifuse/internal/domain LLMService

// LLMService defines the interface for LLM operations
type LLMService interface {
	// StreamChat sends a chat request and streams the response
	StreamChat(ctx context.Context, req *LLMChatRequest, onEvent func(LLMChatEvent) error) error
}

File: console/src/services/api/llm.ts
// LLM Chat API with SSE Streaming

export interface LLMMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface LLMTool {
  name: string
  description: string
  input_schema: object
}

export interface LLMChatRequest {
  workspace_id: string
  integration_id: string
  messages: LLMMessage[]
  max_tokens?: number
  system_prompt?: string
  tools?: LLMTool[]
}

export interface LLMChatEvent {
  type: 'text' | 'tool_use' | 'server_tool_start' | 'server_tool_result' | 'done' | 'error'
  content?: string
  error?: string
  tool_name?: string
  tool_input?: Record<string, unknown>
  input_tokens?: number
  output_tokens?: number
  input_cost?: number   // USD
  output_cost?: number  // USD
  total_cost?: number   // USD
  model?: string
}

export interface StreamChatOptions {
  signal?: AbortSignal
}

export const llmApi = {
  /**
   * Stream chat completion from the LLM API
   * Uses Server-Sent Events (SSE) for real-time streaming
   *
   * @param params - The chat request parameters
   * @param onEvent - Callback for each streamed event
   * @param onError - Optional error callback
   * @param options - Optional options including AbortSignal for cancellation
   *
   * @example
   * const controller = new AbortController()
   * llmApi.streamChat(
   *   { workspace_id: 'ws1', integration_id: 'int1', messages: [{role: 'user', content: 'Hello'}] },
   *   (event) => {
   *     if (event.type === 'text') console.log(event.content)
   *   },
   *   (error) => console.error(error),
   *   { signal: controller.signal }
   * )
   * // To cancel: controller.abort()
   */
  streamChat: async (
    params: LLMChatRequest,
    onEvent: (event: LLMChatEvent) => void,
    onError?: (error: Error) => void,
    options?: StreamChatOptions
  ): Promise<void> => {
    const authToken = localStorage.getItem('auth_token')

    let defaultOrigin = window.location.origin
    if (defaultOrigin.includes('notifusedev.com')) {
      defaultOrigin = 'https://localapi.notifuse.com:4000'
    }
    const apiEndpoint = window.API_ENDPOINT?.trim() || defaultOrigin

    try {
      const response = await fetch(`${apiEndpoint}/api/llm.chat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          ...(authToken ? { Authorization: `Bearer ${authToken}` } : {})
        },
        body: JSON.stringify(params),
        signal: options?.signal
      })

      if (!response.ok) {
        const errorData = await response.json().catch(() => null)
        throw new Error(errorData?.error || `HTTP error: ${response.status}`)
      }

      if (!response.body) {
        throw new Error('No response body')
      }

      const reader = response.body.getReader()
      const decoder = new TextDecoder()
      let buffer = ''

      try {
        while (true) {
          const { done, value } = await reader.read()
          if (done) break

          buffer += decoder.decode(value, { stream: true })
          const lines = buffer.split('\n')
          buffer = lines.pop() || ''

          for (const line of lines) {
            if (line.startsWith('data: ')) {
              try {
                const event: LLMChatEvent = JSON.parse(line.slice(6))
                onEvent(event)

                if (event.type === 'error' && onError) {
                  onError(new Error(event.error || 'Unknown error'))
                }
              } catch {
                // Ignore JSON parse errors for incomplete data
              }
            }
          }
        }
      } finally {
        reader.releaseLock()
      }
    } catch (error) {
      // Don't report AbortError as an error - it's expected when cancelling
      if (error instanceof Error && error.name === 'AbortError') {
        return
      }
      if (onError) {
        onError(error instanceof Error ? error : new Error(String(error)))
      }
    }
  }
}
